goagen/gen_controller/generator.go

```diff
@@ -0,0 +1,329 @@
+	g := &Generator{}
+
+	for _, option := range options {
+		option(g)
+	}
+
+	return g
+}
+
+// Generator is the application code generator.
+type Generator struct {
+	API       *design.APIDefinition // The API definition
+	OutDir    string                // Path to output directory
+	DesignPkg string                // Path to design package, only used to mark generated files.
+	Target    string                // Name of generated "app" package
+	Force     bool                  // Whether to override existing files
+	Pkg       string                // Name of the generated package
+	Resource  string                // Name of the generated file
+	genfiles  []string              // Generated files
+}
+
+// Generate is the generator entry point called by the meta generator.
+func Generate() (files []string, err error) {
+	var (
+		outDir, designPkg, target, ver, res, pkgName string
+		force                                        bool
+	)
+
+	set := flag.NewFlagSet("controller", flag.PanicOnError)
+	set.StringVar(&outDir, "out", "", "")
+	set.StringVar(&designPkg, "design", "", "")
+	set.StringVar(&target, "pkg", "app", "")
+	set.StringVar(&pkgName, "pkg-name", "controller", "")
+	set.StringVar(&res, "res", "controller", "")
+	set.StringVar(&ver, "version", "", "")
+	set.BoolVar(&force, "force", false, "")
+	set.Bool("notest", false, "")
+	set.Parse(os.Args[1:])
+
+	if err := codegen.CheckVersion(ver); err != nil {
+		return nil, err
+	}
+
+	target = codegen.Goify(target, false)
+	g := &Generator{OutDir: outDir, DesignPkg: designPkg, Target: target, Force: force, API: design.Design, Pkg: pkgName, Resource: res}
+
+	return g.Generate()
+}
+
+// Generate produces the skeleton controller service factory.
+func (g *Generator) Generate() (_ []string, err error) {
+	go utils.Catch(nil, func() { g.Cleanup() })
+
+	defer func() {
+		if err != nil {
+			g.Cleanup()
+		}
+	}()
+
+	if g.Target == "" {
+		g.Target = "app"
+	}
+
+	codegen.Reserved[g.Target] = true
+
+	controllerServiceFile := filepath.Join(g.OutDir, g.Resource+".go")
+	if g.Force {
+		os.Remove(controllerServiceFile)
+	}
+	funcs := template.FuncMap{
+		"tempvar":   tempvar,
+		"okResp":    g.okResp,
+		"targetPkg": func() string { return g.Target },
+	}
+	imp, err := codegen.PackagePath(g.OutDir)
+	if err != nil {
+		return nil, err
+	}
+	imp = path.Join(filepath.ToSlash(imp), "app")
+	_, err = os.Stat(controllerServiceFile)
+	if err != nil {
+		// ensure that the output directory exists before creating a new controller
+		if err := os.MkdirAll(g.OutDir, 0755); err != nil {
+			return nil, err
+		}
+		if err = g.createControllerServiceFile(controllerServiceFile, funcs); err != nil {
+			return nil, err
+		}
+	}
+	imports := []*codegen.ImportSpec{
+		codegen.SimpleImport("io"),
+		codegen.SimpleImport("github.com/goadesign/goa"),
+		codegen.SimpleImport(imp),
+		codegen.SimpleImport("golang.org/x/net/websocket"),
+	}
+	err = g.API.IterateResources(func(r *design.ResourceDefinition) error {
```

Comment:
I don't understand this code starting from here to the end of the file. The idea is to generate the the controller code for the one resource, why are we looping through all resources?

Also this should reuse the template and code from the `genmain` package so that when the `genmain` package changes how it generates code this generator output is also updated.